### DISCUSSION TITLE
Output arrays and instances. Requirement of `input`.

### DISCUSSION THREAD

**Discussion by Aymericr** [RobinTail/express-zod-api#712] [Q&A] [answered]
I get type error while declaring `output` from `defaultEndpointsFactory` build method. I couldn't find any answer in the documentation.

```
"typescript": "^4.7.4",
"express-zod-api": "^8.3.4",
```

<img width="587" alt="CleanShot 2022-11-22 at 22 20 27@2x" src="https://user-images.githubusercontent.com/4444492/203466040-7cd55a1a-5b4e-4044-8dc6-1d2409fa29d4.png">

```typescript
// imports
import { Feature, FeatureService } from '@entities/feature.entity';
import { 
  createServer, 
  Routing, 
  z, 
  defaultEndpointsFactory,
  createResultHandler,
  createApiResponse,
  IOSchema,
  EndpointsFactory,
  createConfig,
} from "express-zod-api";

// code
const getFeatures = defaultEndpointsFactory.build({
  method: "get",
  output: z.any().array(),
  handler: async () => {
    const features = new FeatureService();
    const featureList = await features.list();
    return featureList;
  },
});
```
<img width="817" alt="CleanShot 2022-11-22 at 22 43 25@2x" src="https://user-images.githubusercontent.com/4444492/203466061-627aa724-b9a2-48ae-9cf7-1b10846819d2.png">

The type error: `(property) output: z.ZodArray<z.ZodAny, "many">`
```
Type 'ZodArray<ZodAny, "many">' is not assignable to type 'IOSchema<any>'.
  Property 'innerType' is missing in type 'ZodArray<ZodAny, "many">' but required in type 'ZodEffects<ZodObject<any, any, ZodTypeAny, { [x: string]: any; }, { [x: string]: any; }>, { [x: string]: any; }, { [x: string]: any; }>'.ts(2322)
types.d.ts(631, 5): 'innerType' is declared here.
endpoints-factory.d.ts(14, 5): The expected type comes from property 'output' which is declared here on type 'BuildProps<IOSchema<any>, IOSchema<any>, null, {}, "get", string, string>'
```

**Comment by Aymericr** [comment]
This will work though:
<img width="670" alt="CleanShot 2022-11-22 at 23 16 43@2x" src="https://user-images.githubusercontent.com/4444492/203468379-7776aa9e-4292-40f2-ba8e-7b9ffd3f883b.png">
```typescript
const getFeatures = defaultEndpointsFactory.build({
  method: "get",
  input: z.object({}),
  output: z.object({
    features: z.any().array(),
  }),
  handler: async () => {
    const features = new FeatureService();
    const featureList = await features.list();
    return {
      features: featureList
    };
  },
});
```
I would like to be able to write `output: z.instanceof(Feature).array()` without having to nest it under `z.object({...})`

Another weird behavior: why `input` is required for a "get" method?

**Comment by RobinTail** [comment]
Hello @Aymericr .

Thank you for reaching out.

## Regarding arrays

This is not a bug, it's intentionally made that way to encourage developers to always operate objects, as a best practice, since objects can evolve preserving the backward compatibility.
However, if you still want for some reason to respond with an array, you'd need to make a custom ResultHandler.
Find out more here:
https://github.com/RobinTail/express-zod-api/discussions/361
Long story short: your endpoint still have to output an object, but your custom ResultHandler may pick a certain field containing the array and respond with it.

## Regarding instanses

> I would like to be able to write `output: z.instanceof(Feature).array()`

Currently this is not possible, since `z.instanceOf()` means an instance of a class. Classes have to be serialized to something compatible for HTTP transfer (read: JSON-compatible), so in order to avoid uncertainty on the way of serialization, the library does not support usage of `z.instanceOf` in I/O (at least currently). You need to define the schema of the class serialization and do that serialization in the endpoint handler.

## Regarding inputs

> Another weird behavior: why input is required for a "get" method?

If your endpoint does not have any inputs, you need to write `input: z.object({})`, this is required (at least currently) since `input` is always an object produced by the combination of `request` properties (such as `body`, `query` and `params`).
But I agree, that this could be made optional as the future improvement to imply empty object as the default input schema.

I hope I managed to answer all your questions.
Please let me know if something is not clear yet.

**Comment by RobinTail** [comment, marked as answer]
Now supporting array response: https://github.com/RobinTail/express-zod-api/discussions/361#discussioncomment-6879296
(at your own risk)
